numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -123,6 +124,40 @@ legacy_dtype_default_new(PyArray_DTypeMeta *self,
 }
 
 
+static PyObject *
+string_unicode_new(PyArray_DTypeMeta *self, PyObject *args,
+                   PyObject *kwargs)
+{
+  npy_intp size;
+
+  static char *kwlist[] = {"size", NULL};
+
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O&", kwlist,
+               PyArray_IntpFromPyIntConverter, &size)) {
+    return NULL;
+  }
+
+  PyArray_Descr *res = PyArray_DescrNewFromType(self->type_num);
+
+  if (res == NULL) {
+    return NULL;
+  }
+
+  if (size > NPY_MAX_INT) {
+    PyErr_SetString(PyExc_TypeError,
+                    "Characters too large to store inside array.");
+  }
+
+  if (self->type_num == NPY_UNICODE) {
+    // unicode strings are 4 bytes per character
+    size *= 4;
```

Comment:
This could overflow (I admit, this problem is in more places including below, although a bit less pronounced).